seller 만 상품 등록할 수 있게 인가

tableMinPark opened the pull request:
## 개요
- related issue : #83 
- SecurityConfig 에 상품 생성 부분 인증해야 접근할 수 있도록 했습니다.
- Security 에서 인증 관련을 모두 처리하기 때문에 PrincipalDetails가 항상 존재하는 가정하에 구현
- ItemService#createItem 에서 회원 role을 확인하여 SELLER인 경우에만 상품을 등록할 수 있도록 했습니다.

tableMinPark committed:
feat : ItemController#createItem 수정

- userId 를 PrincipalDetails 에서 가져와서 사용하도록 수정
- CreateItemRequest 에 userId 항목 삭제
- SecurityConfig 에 상품 생성 인증 여부 필터링 추가

tableMinPark committed:
feat : ItemController#createItem 수정

- principalDetails 에서 role 받아서 service 넘기는 로직 생성

tableMinPark committed:
feat : ItemService#createItem 수정

- role 이 SELLER 인지 확인하는 로직 생성
- seller 만 상품 등록할 수 있도록 수정

tableMinPark committed:
feat : ExceptionCode 수정

- NO_USER_TO_CREATE_ITEM (회원이 아닐 때 상품을 생성할 수 없는 경우) 에서 NO_ROLE_CREATE_ITEM (SELLER가 아닌 경우) 로 수정

tableMinPark committed:
test : ItemControllerTest 상품 생성 테스트 수정

- ItemControllerTest#createItemNoUserId 삭제 (회원 ID가 없는 경우는 없다고 가정)
- PrincipalDetails 를 SecurityContextHolder 에 추가하는 메서드 ItemControllerTest#addPrincipalDetails 생성
- PrincipalDetails 에서 회원 정보를 가져와 테스트하는 로직으로 테스트 코드 수정

tableMinPark committed:
test : ItemServiceTest 상품 생성 테스트 수정

- User 객체 생성 Builder 패턴으로 수정
- ItemServiceTest#createItemNotJoinUserId 삭제
- ItemServiceTest#createItemNotSeller 생성 (SELLER 가 아닌 회원의 경우 상품 등록할 수 없는 테스트)

daengdaengLee commented on src/main/java/hello/until/item/dto/request/CreateItemRequest.java:
상품 생성 요청 Body 에서 userId 를 받는 건 그대로 두고,
PrincipalDetails 의 유저와 같은지 검증하고,
Body 의 userId 에 해당하는 유저가 SELLER 인지 확인하는 건 어떨까요?

tableMinPark replied:
> 상품 생성 요청 Body 에서 userId 를 받는 건 그대로 두고,
PrincipalDetails 의 유저와 같은지 검증하고,

userId 를 검증하는 것은 좋을 것 같아서 수정했습니다 :)

> Body 의 userId 에 해당하는 유저가 SELLER 인지 확인하는 건 어떨까요?

Security 에서 조회해서 넘기기 때문에 한번 더 조회하지 않아도 될 것 같아서 PrincipalDetails 의 Role 과 비교하는 것도 괜찮지 않을까요?

daengdaengLee replied:
음.. 그래도 되긴 할 것 같아요.
제 의도는 특정 API 를 처리하는 데 필요한 데이터는 해당 요청에 있는 정보가 기반인 게 좋갔다는 거였어요.
Authorization 헤더로 오는 토큰은 요청 정보에 있는 데이터가 유효한지만 검사하고요.
그런데 여기서는 같은 게 확실하니까 인증 유저 객체를 써도 될 것 같네요.

tableMinPark replied:
네 알겠습니다 :)

daengdaengLee reviewed (dismissed):
LGTM!
코멘트 확인 부탁드려요~

tableMinPark committed:
feat : 상품 생성 수정

- Service 로 user 객체를 전달하는 방식으로 수정
- CreateItemRequest 에서 userId 를 받는 것은 동일하게 유지하고, Principal 에서의 userId 와 일치하는지 여부를 확인하는 로직 추가 및 에러 코드 추가

tableMinPark committed:
test : 상품 생성 테스트 수정

- request 로 받은 userId 와 Principal 의 userId 와 일치하는 여부 테스트 추가

dudwns9280 reviewed (dismissed):
좋네용 conflict만 해결하고 머지하면 될 거 같습니다!

tableMinPark committed:
Merge branch 'develop' into feat/83

tableMinPark committed:
test : commit 충돌 코드 수정
